File: app/src/main/java/com/example/alunos/batalhanavalandroid/Navio.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.example.alunos.batalhanavalandroid;

import java.io.Serializable;
import java.util.Random;

/**
 *
 * @author Alunos
 */
public class Navio implements Serializable{

    private int tamanho;
    private int[] posInicial, posFinal;
    private boolean vertical;

    Navio(int tamanho) {
        this.tamanho = tamanho;
        this.posFinal = new int[]{0, 0};
        this.posInicial = new int[]{0, 0};
        this.gerarPosicoes();
    }

    public void gerarPosicoes() {
        Random r = new Random();
        this.vertical = r.nextBoolean();
        this.posInicial[0] = r.nextInt(7);
        this.posInicial[1] = r.nextInt(7);
        if (!this.vertical) {
            this.posFinal[1] = this.posInicial[1];
            this.posFinal[0] = this.posInicial[0] + this.tamanho -1;
            if(this.posFinal[0] >= 7){
                this.posFinal[0] = this.posInicial[0] - this.tamanho + 1;
            }
        } else {
            this.posFinal[0] = this.posInicial[0];
            this.posFinal[1] = this.posInicial[1] + this.tamanho -1;
            if(this.posFinal[1] >= 7){
                this.posFinal[1] = this.posInicial[1] - this.tamanho + 1;
            }
        }
    }

    public int[][] getPosicoes() {
        int[][] posicoes = new int[this.tamanho][2];
        int inicio;

        if(!this.vertical){
            if(this.posInicial[0] < this.posFinal[0]){
                inicio = this.posInicial[0];
            }
            else{
                inicio = this.posFinal[0];
            }

            for(int i = 0; i < this.tamanho; i++){
                posicoes[i][1] = this.posInicial[1];
                posicoes[i][0] = inicio;
                inicio++;
            }
        }
        else{
            if(this.posInicial[1] < this.posFinal[1]){
                inicio = this.posInicial[1];
            }
            else{
                inicio = this.posFinal[1];
            }

            for(int i = 0; i < this.tamanho; i++){
                posicoes[i][0] = this.posInicial[0];
                posicoes[i][1] = inicio;
                inicio++;
            }
        }
        return posicoes;
    }

    public boolean getVertical() {
        return this.vertical;
    }

    public int getTamanho() {
        return this.tamanho;
    }

    public int[] getPosInicial() {
        return this.posInicial;
    }

    public int[] getPosFinal() {
        return this.posFinal;
    }

    public void setVertical(boolean novoVertical) {
        this.vertical = novoVertical;
    }

    public void setTamanho(int novoTamanho) {
        this.tamanho = novoTamanho;
    }

    public void setPosInicial(int[] novaPosInicial) {
        this.posInicial = novaPosInicial;
    }

    public void setPosFinal(int[] novaPosFinal) {
        this.posFinal = novaPosFinal;
    }
}

File: app/src/main/java/com/example/alunos/batalhanavalandroid/Navio.kt
package com.example.alunos.batalhanavalandroid

import java.io.Serializable
import java.util.Random

class Navio internal constructor(var tamanho: Int) : Serializable {

    var posicoes = Array(this.tamanho) { IntArray(2) }
    var vertical = false

    init {
        gerarPosicoesRandomicas(this.tamanho, this.tamanho)
    }

    fun gerarPosicoes(xi: Int, yi: Int, xMax: Int, yMax: Int){
        var inicio: Int

        if (!this.vertical) {

            if (xi + this.tamanho - 1 > xMax) {
                inicio = xi - this.tamanho + 1
            }
            else{
                inicio = xi
            }
            for (i in 0..this.tamanho-1) {
                this.posicoes[i][1] = yi
                this.posicoes[i][0] = inicio
                inicio++
            }

        } else {

            if (yi + this.tamanho - 1 > yMax) {
                inicio = yi - this.tamanho + 1
            }
            else{
                inicio = yi
            }
            for (i in 0..this.tamanho-1) {
                this.posicoes[i][0] = xi
                this.posicoes[i][1] = inicio
                inicio++
            }

        }

    }

    fun gerarPosicoesRandomicas(xMax: Int, yMax: Int){
        val r = Random()
        val xi = r.nextInt(xMax + 1)
        val yi = r.nextInt(yMax + 1)
        this.vertical = r.nextBoolean()
        gerarPosicoes(xi, yi, xMax, yMax)
    }

    fun overlap(navio: Navio): Boolean {
        val nt1: Double
        val dt1: Double
        val nt2: Double
        val dt2: Double
        val t1: Double
        val t2: Double
        var x1: Double
        var y1: Double
        var x2: Double
        var y2: Double
        var x3: Double
        var y3: Double
        var x4: Double
        var y4: Double
        var aux1: Double

        x1 = this.posicoes[0][0].toDouble()
        y1 = this.posicoes[0][1].toDouble()
        x2 = this.posicoes[this.tamanho-1][0].toDouble()
        y2 = this.posicoes[this.tamanho-1][1].toDouble()
        x3 = navio.posicoes[0][0].toDouble()
        y3 = navio.posicoes[0][1].toDouble()
        x4 = navio.posicoes[navio.tamanho-1][0].toDouble()
        y4 = navio.posicoes[navio.tamanho-1][1].toDouble()

        val verticalA = this.vertical
        val verticalB = navio.vertical

        if (verticalA && verticalB && x1 == x3) {
            if (y1 > y2) {
                aux1 = y2
                y2 = y1
                y1 = aux1
            }
            if (y3 > y4) {
                aux1 = y4
                y4 = y3
                y3 = aux1
            }
            return y3 >= y1 && y3 <= y2 || y1 >= y3 && y1 <= y4
        }

        if (!verticalA && !verticalB && y1 == y3) {

            if (x1 > x2) {
                aux1 = x2
                x2 = x1
                x1 = aux1
            }
            if (x3 > x4) {
                aux1 = x4
                x4 = x3
                x3 = aux1
            }
            return x3 >= x1 && x3 <= x2 || x1 >= x3 && x1 <= x4
        }

        /*Aqui eu utilizo uma fórmula para ver se existe intersecção entre dois
        segmentos de reta.
        fonte: http://www.cs.swan.ac.uk/~cssimon/line_intersection.html */

        nt1 = (y3 - y4) * (x1 - x3) + (x4 - x3) * (y1 - y3)
        dt1 = (x4 - x3) * (y1 - y2) - (x1 - x2) * (y4 - y3)
        t1 = nt1 / dt1
        nt2 = (y1 - y2) * (x1 - x3) + (x2 - x1) * (y1 - y3)
        dt2 = (x4 - x3) * (y1 - y2) - (x1 - x2) * (y4 - y3)
        t2 = nt2 / dt2
        return t1 >= 0 && t1 <= 1 && t2 >= 0 && t2 <= 1
    }

    fun posParte(x: Int, y: Int): Boolean {
        var xi = this.posicoes[0][0]
        var yi = this.posicoes[0][1]
        var xf = this.posicoes[this.tamanho-1][0]
        var yf = this.posicoes[this.tamanho-1][1]
        var aux: Int
        val vertical = this.vertical

        if (vertical && x == xi) {
            if (yi > yf) {
                aux = yf
                yf = yi
                yi = aux
            }
            return y >= yi && y <= yf
        }
        else if(!vertical && y == yi){
            if (xi > xf) {
                aux = xf
                xf = xi
                xi = aux
            }
            return x >= xi && x <= xf
        }
        else{
            return false
        }

    }

}